src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java

```diff
@@ -4243,4 +4243,42 @@ public BufferedImage getMascotImage() {
             throw new RandomizerIOException(e);
         }
     }
+
+    @Override
+    protected void randomizeHeldItems(Trainer t, Settings settings) {
+        // This function is called when we know the trainer should have their pokemon given items.
+        t.poketype |= 2; // set the "have item flag"
+
+        List<Integer> toChooseFrom = Gen4Constants.allHeldItems;
+        if (settings.getRandomItemsType() == Settings.TrainersItemsMod.CONSUMABLE_ONLY) {
+            toChooseFrom = Gen4Constants.consumableHeldItems;
+        }
+        if (settings.getRandomItemsFor() == Settings.TrainersItemsFor.ALL_POKEMON) {
+            for (TrainerPokemon tp : t.pokemon) {
+                tp.heldItem = toChooseFrom.get(random.nextInt(toChooseFrom.size()));
+            }
+        } else {
+            int maxLevel = -1;
+            TrainerPokemon highestLevelPoke = null;
+            for (TrainerPokemon tp : t.pokemon) {
+                if (tp.level > maxLevel) {
+                    highestLevelPoke = tp;
+                    maxLevel = tp.level;
+                }
+                tp.heldItem = 0; // set every poke to no item (just in case there was one originally).
+            }
+            if (highestLevelPoke == null) {
+                return; // should never happen - trainer had zero pokes
+            }
+            highestLevelPoke.heldItem = toChooseFrom.get(random.nextInt(toChooseFrom.size()));
+        }
+    }
+
+    private boolean trainersPokemonHaveItems(Trainer tr) {
+        return (tr.poketype & 2) == 2;
+    }

```

Comment:
This seems to work the exact same way in every generation, so this would work better as a method in the `Trainer` class. (Same with the "custom moves" one below.)